### ISSUE TITLE
Add shorthand to pack timestamp

### ISSUE THREAD

**Issue by gappleto97** [msgpack/msgpack-c#638]
Since this is now part of the msgpack spec, shouldn't there be an easy way to pack it in C/C++?

**Comment by redboltz** [comment]
It's nice to have. I think that `std::chrono` is good target for mapping.

**Comment by ChristThePhone** [comment]
When is this one going to be pulled into the master ?

**Comment by redboltz** [comment]
I'm implementing timestamp support.

Here is developing branch:
https://github.com/redboltz/msgpack-c/tree/fix_638

Adaptor is implemented but tests are not yet but you can try it.

Unfortunately, I'm tied up with my company's project. I can't promise when.

**Comment by ChristThePhone** [comment]
Thanks, I already had that branch downloaded and found a compile error trying to use it. I'm a total github newbie not being familiar with how to collaborate so in case there's anything I can help with to get this patch out of the door please let me now.

**Comment by cmpute** [comment]
Any progress on this? It's quite useful in my project!

**Comment by gappleto97** [comment]
Can we also get support on the C version? That was what I was most interested in.

**Comment by redboltz** [comment]
Here is update implementation:
https://github.com/redboltz/msgpack-c/tree/fix_638_2

Some tests fails.
I'm not sure `std::chrono` can explain msgpack timestamp resolution
https://github.com/msgpack/msgpack/blob/master/spec.md#timestamp-extension-type

```
Running main() from gtest_main.cc
[==========] Running 60 tests from 32 test cases.
[----------] Global test environment set-up.
[----------] 14 tests from MSGPACK_CPP11
[ RUN      ] MSGPACK_CPP11.simple_tuple
[       OK ] MSGPACK_CPP11.simple_tuple (0 ms)

...

[----------] 7 tests from MSGPACK_CHRONO
[ RUN      ] MSGPACK_CHRONO.system_clock
[       OK ] MSGPACK_CHRONO.system_clock (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_32
[       OK ] MSGPACK_CHRONO.system_clock_32 (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_32_max
[       OK ] MSGPACK_CHRONO.system_clock_32_max (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_64
[       OK ] MSGPACK_CHRONO.system_clock_64 (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_64_max
/home/kondo/work/msgpack-c/test/msgpack_cpp11.cpp:941: Failure
Value of: 0
Expected: memcmp(sbuf.data(), packed, sizeof(packed))
Which is: -213
[  FAILED  ] MSGPACK_CHRONO.system_clock_64_max (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_96
[       OK ] MSGPACK_CHRONO.system_clock_96 (0 ms)
[ RUN      ] MSGPACK_CHRONO.system_clock_96_max
/home/kondo/work/msgpack-c/test/msgpack_cpp11.cpp:982: Failure
Value of: 0
Expected: memcmp(sbuf.data(), packed, sizeof(packed))
Which is: 16
[  FAILED  ] MSGPACK_CHRONO.system_clock_96_max (0 ms)
[----------] 7 tests from MSGPACK_CHRONO (0 ms total)

[----------] Global test environment tear-down
[==========] 60 tests from 32 test cases ran. (7179 ms total)
[  PASSED  ] 58 tests.
[  FAILED  ] 2 tests, listed below:
[  FAILED  ] MSGPACK_CHRONO.system_clock_64_max
[  FAILED  ] MSGPACK_CHRONO.system_clock_96_max

 2 FAILED TESTS
```

**Comment by redboltz** [comment]
@gappleto97 , C version doesn't have converting functionality. It's not only TIMESTAMP types but also other all types.

I have no plan to expand msgpack_object (https://github.com/redboltz/msgpack-c/blob/master/include/msgpack/object.h).

What kind of support do you expect?

Something like this?

```c
bool convert_to_tstype(msgpack_object const* obj, tstype* ts);
```

If it does, is there any existing potable tstype? Or msgpack should introduce the new type?

**Comment by gappleto97** [comment]
I don't know of any nanosecond precision functions that are portable, so it would seem like msgpack needs to define that. There are a bunch that seem to work on Unix systems, though. `struct timespec` would probably get you the most coverage on that front, and rewriting parts of it for Windows systems doesn't look like it would be *too* painful. [This stackoverflow post](https://stackoverflow.com/a/44136515/4748474) covers a way to do it on Windows systems.

I would recommend making a msgpack defined alias of `struct timespec`, and then writing a function for Windows developers to retrieve it.

**Comment by gappleto97** [comment]
Added C packing support for your fork in https://github.com/redboltz/msgpack-c/pull/1

Hopefully this design seems reasonable. If not, I am happy to work with you on it. It uses portable (C11) functions. The only thing I didn't see was how to run the tests for this repo, so I have not run them on that code yet. I do know that it compiles though.

**Comment by redboltz** [comment]
I implemented #706 . It is the C++ version of TimeStamp support. Finally, all test has been passed.

**Comment by redboltz** [comment]
@gappleto97 , thank you for sending the PR. 

I think that the type `timespec` is the type for C11 and C++17. Maybe we need to support older version.

Maybe some type should be defined as follows:

```c
#if /* C11 */
typedef timespec msgpack_timestamp;
#else
struct msgpack_timestamp {
    int64_t sec;
    uint32_t nanosec;
};
#endif
```

As I mentioned https://github.com/msgpack/msgpack-c/issues/638#issuecomment-401542803, I don't want to expand msgpack_object. Because if new other MessagePack types will be added in the future using EXT with minus type value, the type need to added an union member of the msgpack_object. And its size is unpredictable. I want to avoid size bloat. 
So for TimeStamp, msgpack_object is keep EXT with minus type (-1), and providing converting function.

```c
bool convert_to_tstype(msgpack_object const* obj, msgpack_timestamp* ts);
```

**Comment by redboltz** [comment]
I implemented C version as #707. This is the first minimal support for C.
It only defines `msgpack_timestamp`.

```c
typedef struct msgpack_timestamp {
    int64_t tv_sec;
    uint32_t tv_nsec;
} msgpack_timestamp;
```

C11 and C++17 's  `timespec` support is not implemented. Conditional type definition is pretty complected. In addition, the member `tv_sec`  of the `timespec` has the type `time_t` and AFAIK the actual type of `time_t` is implementation defined.

I decided providing `msgpack_timestamp` that has enough resolution to represent MssagePack TimeStamp and if someone needs to use `timespec`, convert from/to `msgpack_timestamp` and `timespec`. It's easy.
